### PR TITLE
Er-965_Validate reference data

### PR DESCRIPTION
**Pull request by shomavg**
Validation mainly is mainly on Bankholidays, Standards and Frameworks.

### CODE REVIEW

File: src/Shared/Recruit.Vacancies.Client/Infrastructure/ReferenceData/BankHolidays/BankHolidayUpdateService.cs [hunk 1]
```diff
@@ -1,48 +1,67 @@
...
+            if (await ValidateBankHolidayData(bankHolidaysFromApi))
+            {
+                await _referenceDataWriter.UpsertReferenceData(bankHolidaysFromApi);
+                _logger.LogInformation($"Upserted bank holidays into ReferenceData store. Last England and Wales date:{bankHolidaysFromApi.Data.EnglandAndWales.Events.Last().Date}");
+            }
+            _logger.LogInformation("ReferenceData not updated as the ETag value is the same.");
+        }
 
-            _logger.LogInformation($"Upserted bank holidays into ReferenceData store. Last England and Wales date:{bankHolidays.Data.EnglandAndWales.Events.Last().Date}");
+        private async Task<bool> ValidateBankHolidayData(BankHolidays bankHolidaysFromApi)
```
Comment (chrisjensenuk) [inline src/Shared/Recruit.Vacancies.Client/Infrastructure/ReferenceData/BankHolidays/BankHolidayUpdateService.cs]: Can we rename this to `HasBankHolidayDataChanged`

File: src/Shared/Recruit.Vacancies.Client/Infrastructure/ReferenceData/BankHolidays/BankHolidays.cs [hunk 2]
```diff
@@ -1,21 +1,39 @@
...
 using System.Collections.Generic;
+using System.IO;
+using System.Runtime.Serialization.Json;
+using System.Text;
+using Microsoft.Net.Http.Headers;
 using MongoDB.Bson.Serialization.Attributes;
 
 namespace Esfa.Recruit.Vacancies.Client.Infrastructure.ReferenceData.BankHolidays
 {
+    public static class MySerializer
```
Comment (chrisjensenuk) [inline src/Shared/Recruit.Vacancies.Client/Infrastructure/ReferenceData/BankHolidays/BankHolidays.cs]: Do we need this class?

File: src/Shared/Recruit.Vacancies.Client/Infrastructure/ReferenceData/BankHolidays/BankHolidayProvider.cs [hunk 3]
```diff
@@ -36,5 +33,10 @@ public async Task<List<DateTime>> GetBankHolidaysAsync()
                         .ToList();
                 });
         }
+
+        public async Task<BankHolidays> GetBankHolidayListAsync()
```
Comment (chrisjensenuk) [inline src/Shared/Recruit.Vacancies.Client/Infrastructure/ReferenceData/BankHolidays/BankHolidayProvider.cs]: I'd delete this from here and just talk directly to referenceDateReader from BankHolidayUpdateService

File: src/Shared/Recruit.Vacancies.Client/Infrastructure/ReferenceData/BankHolidays/BankHolidayUpdateService.cs [hunk 4]
```diff
@@ -1,48 +1,67 @@
...
 
-namespace Esfa.Recruit.Vacancies.Client.Infrastructure.Services
+namespace Esfa.Recruit.Vacancies.Client.Infrastructure.ReferenceData.BankHolidays
 {
     public class BankHolidayUpdateService : IBankHolidayUpdateService
     {
         private readonly BankHolidayConfiguration _config;
         private readonly IReferenceDataWriter _referenceDataWriter;
         private readonly ILogger<BankHolidayUpdateService> _logger;
+        private readonly IBankHolidayProvider _bankHolidayProvider;
```
Comment (chrisjensenuk) [inline src/Shared/Recruit.Vacancies.Client/Infrastructure/ReferenceData/BankHolidays/BankHolidayUpdateService.cs]: I'd remove this reference to the `IBankHolidayProvider` and just add a reference to `IReferenceDataReader ` and get the bank holidays directly.

File: src/Shared/Recruit.Vacancies.Client/Infrastructure/ReferenceData/BankHolidays/BankHolidayUpdateService.cs [hunk 5]
```diff
@@ -1,48 +1,67 @@
...
+        }
 
-            _logger.LogInformation($"Upserted bank holidays into ReferenceData store. Last England and Wales date:{bankHolidays.Data.EnglandAndWales.Events.Last().Date}");
+        private async Task<bool> ValidateBankHolidayData(BankHolidays bankHolidaysFromApi)
+        {
+            var bankHolidaysFromDb = await _bankHolidayProvider.GetBankHolidayListAsync();
+            var bankHolidaysFromApiJson = JsonConvert.SerializeObject(bankHolidaysFromApi.Data);
+            var bankHolidaysFromDbJson = JsonConvert.SerializeObject(bankHolidaysFromDb.Data);
+            var areEqual = JToken.DeepEquals(bankHolidaysFromApiJson, bankHolidaysFromDbJson);
+            if (!areEqual)
```
Comment (chrisjensenuk) [inline src/Shared/Recruit.Vacancies.Client/Infrastructure/ReferenceData/BankHolidays/BankHolidayUpdateService.cs]: `return !areEqual;`

File: src/Shared/Recruit.Vacancies.Client/Infrastructure/ReferenceData/BankHolidays/BankHolidayUpdateService.cs [hunk 6]
```diff
@@ -1,48 +1,67 @@
...
+                Data = response.Data             
             };
 
-            await _referenceDataWriter.UpsertReferenceData(bankHolidays);
+            if (await ValidateBankHolidayData(bankHolidaysFromApi))
+            {
+                await _referenceDataWriter.UpsertReferenceData(bankHolidaysFromApi);
+                _logger.LogInformation($"Upserted bank holidays into ReferenceData store. Last England and Wales date:{bankHolidaysFromApi.Data.EnglandAndWales.Events.Last().Date}");
+            }
+            _logger.LogInformation("ReferenceData not updated as the ETag value is the same.");
```
Comment (chrisjensenuk) [inline src/Shared/Recruit.Vacancies.Client/Infrastructure/ReferenceData/BankHolidays/BankHolidayUpdateService.cs]: Should this be in an `else` or return early.

File: src/Shared/Recruit.Vacancies.Client/Infrastructure/ReferenceData/BankHolidays/BankHolidayUpdateService.cs (hunk 1 above)
Comment (shomavg) [reply src/Shared/Recruit.Vacancies.Client/Infrastructure/ReferenceData/BankHolidays/BankHolidayUpdateService.cs]: The changes have been made, can you please review again?

File: src/Shared/Recruit.Vacancies.Client/Infrastructure/ReferenceData/ApprenticeshipProgrammes/ApprenticeshipProgrammesUpdateService.cs [hunk 7]
```diff
@@ -39,8 +43,8 @@ public async Task UpdateApprenticeshipProgrammesAsync()
             {
                 Task.WaitAll(tasks.ToArray());
 
-                var standardsFromApi = standardsTask.Result.ToList();
-                var frameworksFromApi = frameworksTask.Result.ToList();
+                var standardsFromApi = standardsTask.Result.Distinct().ToList();
```
Comment (sachpatel) [inline src/Shared/Recruit.Vacancies.Client/Infrastructure/ReferenceData/ApprenticeshipProgrammes/ApprenticeshipProgrammesUpdateService.cs]: the distinct calls here conflict with the other distinct call on insertion.

File: src/Shared/Recruit.Vacancies.Client/Infrastructure/ReferenceData/ApprenticeshipProgrammes/ApprenticeshipProgrammesUpdateService.cs [hunk 8]
```diff
@@ -75,6 +77,21 @@ public async Task UpdateApprenticeshipProgrammesAsync()
...
+        private async Task ValidateList(List<ApprenticeshipProgramme> apprenticeshipProgrammesFromApi)
+        {
+            var apprenticeshipProgrammesFromDb = await _programmeProvider.GetApprenticeshipProgrammesAsync(true);
+            var apiCount = apprenticeshipProgrammesFromApi.Count;
+            var dbCount = apprenticeshipProgrammesFromDb.Count();
+            var difference = Math.Abs(apiCount - dbCount);
+            double percent = (double)(difference * 100) / apiCount; 
+            if (percent > AcceptablePercentage)
+            {
+                _logger.LogWarning("There is a difference of more than 10% between the existing ApprenticeshipProgrammes and the " +
```
Comment (sachpatel) [inline src/Shared/Recruit.Vacancies.Client/Infrastructure/ReferenceData/ApprenticeshipProgrammes/ApprenticeshipProgrammesUpdateService.cs]: Can you interpolate the Acceptable percentage in this string please.

File: src/Shared/Recruit.Vacancies.Client/Infrastructure/ReferenceData/BankHolidays/BankHolidayUpdateService.cs (hunk 6 above)
Comment (sachpatel) [reply src/Shared/Recruit.Vacancies.Client/Infrastructure/ReferenceData/BankHolidays/BankHolidayUpdateService.cs]: we are not using Etag so this info log is wrong

File: src/Shared/UnitTests/Vacancies.Client/Infrastructure/ReferenceData/ApprenticeshipProgrammes/ApprenticeshipProgrammesUpdaterServiceTests.cs [hunk 9]
```diff
@@ -16,7 +15,7 @@
 using Xunit;
 using ApprenticeshipProgrammesReferenceData = Esfa.Recruit.Vacancies.Client.Infrastructure.ReferenceData.ApprenticeshipProgrammes.ApprenticeshipProgrammes;
 
-namespace Esfa.Recruit.UnitTests.Vacancies.Client.Infrastructure.ReferenceData.ApprenticeshipProgrammes
+namespace Esfa.Recruit.Vacancies.Client.UnitTests.Vacancies.Client.Infrastructure.ReferenceData.ApprenticeshipProgrammes
```
Comment (sachpatel) [inline src/Shared/UnitTests/Vacancies.Client/Infrastructure/ReferenceData/ApprenticeshipProgrammes/ApprenticeshipProgrammesUpdaterServiceTests.cs]: Can you remove the Vacancies.Client before the UnitTests part, we need to keep these namespaces manageable and short. Having the same two bits multiple times makes this long and difficult.

File: src/Shared/Recruit.Vacancies.Client/Infrastructure/ReferenceData/ApprenticeshipProgrammes/ApprenticeshipProgrammesUpdateService.cs (hunk 7 above)
Comment (shomavg) [reply src/Shared/Recruit.Vacancies.Client/Infrastructure/ReferenceData/ApprenticeshipProgrammes/ApprenticeshipProgrammesUpdateService.cs]: removed it.

File: src/Shared/UnitTests/Vacancies.Client/Infrastructure/ReferenceData/ApprenticeshipProgrammes/ApprenticeshipProgrammesUpdaterServiceTests.cs (hunk 9 above)
Comment (shomavg) [reply src/Shared/UnitTests/Vacancies.Client/Infrastructure/ReferenceData/ApprenticeshipProgrammes/ApprenticeshipProgrammesUpdaterServiceTests.cs]: But shouldnt the namespace match the location of the file?

Comment (sachpatel) [reply src/Shared/UnitTests/Vacancies.Client/Infrastructure/ReferenceData/ApprenticeshipProgrammes/ApprenticeshipProgrammesUpdaterServiceTests.cs]: No who said that? This whole thing needs revisiiting, the tests in that folder have all had namespace renamed to what I suggested. Please turn it back. We can address the rest next week.

Comment (shomavg) [reply src/Shared/UnitTests/Vacancies.Client/Infrastructure/ReferenceData/ApprenticeshipProgrammes/ApprenticeshipProgrammesUpdaterServiceTests.cs]: done.

File: src/Shared/Recruit.Vacancies.Client/Infrastructure/ReferenceData/BankHolidays/BankHolidayUpdateService.cs [hunk 10]
```diff
@@ -2,47 +2,64 @@
...
             };
 
-            await _referenceDataWriter.UpsertReferenceData(bankHolidays);
+            if (await HasBankHolidayDataChanged(bankHolidaysFromApi))
+            {
+                await _referenceDataWriter.UpsertReferenceData(bankHolidaysFromApi);
+                _logger.LogInformation($"Upserted bank holidays into ReferenceData store. Last England and Wales date:{bankHolidaysFromApi.Data.EnglandAndWales.Events.Last().Date}");
+            }
+            else
+                _logger.LogInformation("ReferenceData not updated as their is no change.");
```
Comment (chrisjensenuk) [inline src/Shared/Recruit.Vacancies.Client/Infrastructure/ReferenceData/BankHolidays/BankHolidayUpdateService.cs]: typo `there`

File: src/Shared/Recruit.Vacancies.Client/Recruit.Vacancies.Client.csproj [hunk 11]
```diff
@@ -20,6 +20,7 @@
   </PropertyGroup>
   <ItemGroup>
     <PackageReference Include="HtmlSanitizer" Version="4.0.205" />
+    <PackageReference Include="JsonDiffPatch.Net" Version="2.1.0" />
```
Comment (chrisjensenuk) [inline src/Shared/Recruit.Vacancies.Client/Recruit.Vacancies.Client.csproj]: Don't think we need this

**Pull request merged**